examples/other/integration_pytorch.py

```diff
@@ -0,0 +1,530 @@
+    return model
+
+
+def append_moments(x: torch.Tensor) -> torch.Tensor:
+    second_order_moments = torch.einsum('bi,bj->bij', x, x)
+
+    # Get the upper triangular indices
+    rows, cols = torch.triu_indices(second_order_moments.shape[1], second_order_moments.shape[2])
+
+    # Extract the upper triangular part for each MxM matrix
+    upper_triangular_values = second_order_moments[:, rows, cols]
+
+    appended_x = torch.cat((x, upper_triangular_values.view(x.shape[0], -1)), dim=1)
+    return appended_x
+
+
+class ExampleDataset(Dataset):
+    def __init__(self):
+        super().__init__()
+        file_path = Path(__file__).parents[1] / 'data/retinal.obj'
+        self._mesh = trimesh.load(file_obj=file_path)
+        self._transforms = transforms.Compose([transforms.FaceToEdge(remove_faces=False)])
+
+    def len(self):
+        return 1
+
+    def get(self, idx):
+        # place mesh in canonical form, aligned with its PCA principal directions
+        vertices = self._mesh.vertices.astype(np.float32)
+        vertices = canonical_form(vertices=vertices)
+        faces = self._mesh.faces
+
+        # calculate principal curvatures at each points
+        _, _, pv1, pv2 = igl.principal_curvature(v=vertices, f=faces)
+
+        # calculate gaussian curvature
+        k = pv1 * pv2
+        data = Data(
+            x=torch.from_numpy(vertices),
+            k=torch.from_numpy(k),
+            face=torch.from_numpy(faces).T)
+
+        # infer edges based on face data
+        data = self._transforms(data)
+
+        return data
+
+
+class ExampleDataModule(LightningDataModule):
+    def __init__(self):
+        super().__init__()
+        self._train_dataset = ExampleDataset()
+        self._validation_dataset = ExampleDataset()
+
+    def train_dataloader(self):
+        return DataLoader(self._train_dataset, batch_size=1)
+
+
+class ExampleModule(LightningModule):
+    def __init__(self):
+        super().__init__()
+        convolutions = [
+            torch_geometric.nn.conv.GATv2Conv(9, 16, head=2),
+            torch_geometric.nn.conv.GATv2Conv(16, 32, head=2),
+        ]
+        self._gnn = create_gnn(convolutions=convolutions, activation='sine', batch_norm=False)
+        self._mlp = create_mlp(features=[32, 16, 8, 4, 1], activation='sine', batch_norm=False)
+        self._loss_fn = nn.SmoothL1Loss()
+
+    def configure_optimizers(self):
+        optimizer = optim.AdamW(self.parameters(), lr=1e-2)
+        return optimizer
+
+    def training_step(self, batch: Batch, index: int):
+        data = batch[0]
+        x = append_moments(x=data.x)
+        embeddings = self._gnn(x=x, edge_index=data.edge_index)
+        pred_k = self._mlp(embeddings).squeeze(dim=1)
+        loss = self._loss_fn(pred_k, data.k)
+
+        return {
+            'loss': loss,
+            'pred_k': pred_k,
+            'gt_k': data.k
+        }
+
+
+class PyGfxCallback(Callback):
```

Comment:
I would've separated the logic that initializes and manages the plot, and the callbacks which update it. This is a massive class (in terms of logic) which is doing multiple things, it can be difficult to parse out how it works. The callback class should be handling only callback related logic.

Reply:
Thanks for the review, I will address it in the next commit.

Reply:
Fixed